core/planner.go

```diff
@@ -0,0 +1,158 @@
+package core
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+)
+
+// RatesByPrice implements sort.Interface based on price
+type RatesByPrice []api.Rate
+
+func (a RatesByPrice) Len() int {
+	return len(a)
+}
+
+func (a RatesByPrice) Less(i, j int) bool {
+	if a[i].Price == a[j].Price {
+		return a[i].Start.After(a[j].Start)
+	}
+	return a[i].Price < a[j].Price
+}
+
+func (a RatesByPrice) Swap(i, j int) {
+	a[i], a[j] = a[j], a[i]
+}
+
+type Planner struct {
+	log         *util.Logger
+	tariff      api.Tariff
+	mux         sync.Mutex
+	cheap       float64
+	cheapactive bool
+	last        time.Time
+}
+
+func NewPlanner(log *util.Logger, tariff api.Tariff) *Planner {
+	return &Planner{
+		log:    log,
+		cheap:  0.2, // TODO
+		tariff: tariff,
+	}
+}
+
+func (t *Planner) IsCheap(duration time.Duration, end time.Time) (bool, error) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	if end.Before(time.Now()) {
+		t.cheapactive = false
+		return false, nil
+	}
+
+	duration = time.Duration(float64(duration) * 1.05) // increase by 5%
```

Comment:
warum?

Reply:
Anpassung der geschätzen Zeit an die tatsächliche Zeit. Beim Ziellladen haben wir hier 20%, sollten wir dann vereinheitlichen

Reply:
Denke auch. Das ist eine globale Annahme die const sein sollte

Reply:
Const chargeEfficiency = 0.95

Haben wir ähnlich auch im ChargeTimer- einmal sollte reichen und beide die gleiche Konstante verwenden.

Reply:
ja, sehe ich auch so. Ich kann darauf aber nicht zugreifen, ist in einem anderen package.

Reply:
Einfach „public“ machen- großer Anfangsbuchstabe. Dann kannst Du die auch nutzen.

Reply:
nö, so einfach ist es nicht. Ich müsste vermutlich auch das soc package importieren. Ist auch nicht schön hier eine weitere Abhängigkeit dazu zu bekommen. Ev. sollte es dann wo anders definiert werden?

Reply:
Mach einfach eine lokale Konstante, verschieben können wir sie immer noch.